accepted/install-locations.md

```diff
@@ -0,0 +1,241 @@
+
+* Windows - the installer will store the install location in 32-bit registry (for both 32-bit and 64-bit installations):
+  * `HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation` - where `<arch>` is one of the `x86`, `x64`, `arm` or `arm64`.
+  Note that this registry key is "redirected" that means that 32-bit processes see different copy of the key then 64bit processes. So it's important that both installers and the host access only the 32-bit view of the registry.
+* Linux and macOS - the installer will store the install location in a text file
+  * `/etc/dotnet/install.conf` - The file will be in the `.ini` [format](https://en.wikipedia.org/wiki/INI_file). In it the key `InstallLocation` (in no section) will contain the path to the global location. The host will have a simplistic parser for this file, which will only support empty lines, comment lines (starts with #) and the one key/value pair. Any additional keys will be ignored. The host will trim any spaces around the equal sign and then treat the rest of the line after the equal sign as the install location path (no quotes or escaping).
+
+The muxer is registered on the system in such a way that it is readily accessible without providing full path to it. The exact mechanism is platform specific and outside of the scope of this document. (For example on Windows it's added to `PATH`)
+
+### Private install location
+This allows users to install private copy of .NET Core into any directory. This copy should not be used by anything unless explicitly asked for.
+
+There is not installer for this scenario, these installs are produced by simply copying the right directory structure into a custom location.
+
+If the muxer (`dotnet`) is used then it must be invoked directly from the private install location (it's not registered on `PATH` in any way).  
+If other hosts are used then the private install location must be specified in environment variable:
+* `DOTNET_ROOT`
+* `DOTNET_ROOT(x86)` for 32-bit process running on 64-bit Windows.
+
+*Note: this type of install is supported in all versions of .NET Core and there are no changes proposed to how it's created and used.*
+
+
+## Host behavior for 3.0
+
+### `hostfxr` search
+`dotnet` only looks right next to itself.
+
+All other hosts will search for the first location which exists in this order:
+* the directory where the host resides (this is important mostly for `apphost` as self-contained apps will have the `hostfxr` in the same directory).
+* `DOTNET_ROOT` (or `DOTNET_ROOT(x86)`) environment variable, in the `host/fxr/<version>` subdirectory.
+* :new: global registered install location, in the `/host/fxr/<version>` subdirectory  
+  **We've committed to support this on Windows for .NET Core 3.0**
+  * Windows - `HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation` (32-bit view)  
+  * Linux/macOS - `/etc/dotnet/install.conf`  
```

Comment:
It would be great if we can also have this on Linux for 3.0 (https://github.com/dotnet/designs/issues/64).

Reply:
We can help with implementing this on Linux if it's a question of devel resources.

Reply:
In terms of implementation my current plan is for us to implement it in the hosts first. The installers which install to the default location don't necessarily need to write the file as the hosts will work without that. I would like even the default location to use the config file just so that all cases are consistent, but it's not absolutely necessary, so will probably happen later on.

Obviously if some other distro wanted to install to a non-default location its installers would have to write the file for the system to work.

Reply:
> Obviously if some other distro wanted to install to a non-default location its installers would have to write the file for the system to work.

The default location that is put into the host doesn't match with where source-build Fedora packages place dotnet (https://github.com/dotnet/designs/issues/64). So we would very much like to provide a config file with the 3.0 packages that makes framework dependent applications work without setting DOTNET_ROOT.

Reply:
That's exactly the goal of this proposal. I'll ping you when we have something working... would very much appreciate if you could test it then.

Reply:
+1 we can easily sync (through one of the many Microsoft-RedHat regular channels) and help with Linux specific implementation of just about anything, just give us a nudge if there is something! =)

Reply:
The has been merged and is now part of the latest runtime/SDK packages. You can get the latest here: https://github.com/dotnet/core-sdk  
I verified that downloading the tarball into a custom location and adding the `etc/dotnet/install_location` file with a single line pointing to the custom location makes it work. That is when I create a new app with it and then run the executable directly (the apphost) without setting any environment variables, it runs and it does so against the new custom location.

Can you please give it a try in your environment?
I guess the most interesting case would be to verify that it works when you put it into the desired distro-specific location and add the install location file. And then validate that it is possible to run apps with the executable directly without any other special environment settings.

Reply:
This works.

I've downloaded the sdk and used it to create a FDD linux-x64 application.

Trying to run it gives:

> A fatal error occurred. The required library libhostfxr.so could not be found.
If this is a self-contained application, that library should exist in [/tmp/console2/bin/Debug/netcoreapp3.0/linux-x64/publish/].
If this is a framework-dependent application, install the runtime in the global location [/usr/share/dotnet] or use the DOTNET_ROOT environment variable to specify the runtime location.

Creating `/etc/dotnet/install_location` and adding the path to folder containing `dotnet`.

> Hello World!

Reply:
> If this is a framework-dependent application, install the runtime in the global location [/usr/share/dotnet] or use the DOTNET_ROOT environment variable to specify the runtime location.

Maybe we should update that error message to also point people to `/etc/dotnet/install_location`?

Reply:
@eerhardt Good suggestion - dotnet/core-setup#6546